File: heap_sort/heap_sort.py
import random


def sort(data):
    build_max_heap(data)

    print(data)

    for i in reversed(range(1, len(data))):
        data[0], data[i] = data[i], data[0]
        heapify(data, 0, i - 1)

        print(data)


def build_max_heap(data):
    for i in reversed(range(len(data) // 2)):
        heapify(data, i, len(data) - 1)


def heapify(data, idx_of_target, idx_of_heap):
    idx_of_largest = idx_of_target

    idx_of_left = 2 * idx_of_target + 1
    if idx_of_left <= idx_of_heap and data[idx_of_left] > data[idx_of_largest]:
        idx_of_largest = idx_of_left
    idx_of_right = 2 * (idx_of_target + 1)
    if idx_of_right <= idx_of_heap and data[idx_of_right] > data[idx_of_largest]:
        idx_of_largest = idx_of_right

    if idx_of_largest != idx_of_target:
        data[idx_of_target], data[idx_of_largest] = data[idx_of_largest], data[idx_of_target]

        heapify(data, idx_of_largest, idx_of_heap)


if __name__ == '__main__':
    data = [random.randint(1, 20) for _ in range(20)]

    print("--- before ---")
    print(data)

    print("--- sort ---")
    sort(data)

    print("--- after ---")
    print(data)

File: heap_sort/heap_sort.go
package main

import (
	"fmt"
	"math/rand"
)

func createData() []int {
	data := make([]int, 20)

	for i := 0; i < 20; i++ {
		data[i] = rand.Intn(20)
	}

	return data
}

func sort(data []int) {
	buildMaxHeap(data)

	fmt.Println(data)

	n := len(data) - 1
	for i := n; i >= 1; i-- {
		data[0], data[i] = data[i], data[0]

		fmt.Println(data)

		maxHeapify(data, 0, i-1)
	}
}

func buildMaxHeap(data []int) {
	n := len(data) - 1

	for i := n / 2; i >= 0; i-- {
		maxHeapify(data, i, n)
	}
}

func maxHeapify(data []int, idxOfTarget, idxOfHeap int) {
	idxOfLargest := idxOfTarget
	idxOfLeft := 2*idxOfTarget + 1
	idxOfRight := 2 * (idxOfTarget + 1)

	if idxOfLeft < idxOfHeap && data[idxOfLeft] > data[idxOfLargest] {
		idxOfLargest = idxOfLeft
	}
	if idxOfRight < idxOfHeap && data[idxOfRight] > data[idxOfLargest] {
		idxOfLargest = idxOfRight
	}

	if idxOfLargest != idxOfTarget {
		data[idxOfTarget], data[idxOfLargest] = data[idxOfLargest], data[idxOfTarget]

		maxHeapify(data, idxOfLargest, idxOfHeap)
	}
}

func main() {
	data := createData()

	fmt.Println("--- before ---")
	fmt.Println(data)

	fmt.Println("--- sort ---")
	sort(data)

	fmt.Println("--- after ---")
	fmt.Println(data)
}
